src/main/scala/mesosphere/marathon/tasks/TaskTrackerImpl.scala

```diff
@@ -0,0 +1,216 @@
+      case None =>
+        Future.failed(new IllegalStateException(s"task [$taskId] of app [$appId] does not exist"))
+    }
+  }
+
+  private[this] def processUpdate(app: InternalApp, task: MarathonTask, status: TaskStatus): Future[_] = {
+    def updateTaskOnStateChange(task: MarathonTask): Future[_] = {
+      def statusDidChange(statusA: TaskStatus, statusB: TaskStatus): Boolean = {
+        val healthy = statusB.hasHealthy &&
+          (!statusA.hasHealthy || statusA.getHealthy != statusB.getHealthy)
+
+        healthy || statusA.getState != statusB.getState
+      }
+
+      if (statusDidChange(task.getStatus, status)) {
+        updateTask(task.toBuilder.setStatus(status).build)
+      }
+      else {
+        log.debug(s"Ignoring status update for ${task.getId}. Status did not change.")
+        Future.successful(())
+      }
+    }
+
+    def updateTask(updatedTask: MarathonTask): Future[_] = {
+      val appId = app.appName
+      val taskId: String = updatedTask.getId
+      log.debug(s"updating task [$taskId] of app [$appId]: {}", updatedTask)
+
+      getTaskMap(appId) += (taskId -> updatedTask)
+      repo.store(updatedTask).map(_ => updatedTask)
+    }
+
+    status.getState match {
+      case TASK_ERROR | TASK_FAILED | TASK_FINISHED | TASK_KILLED | TASK_LOST =>
+        removeTask(app, task.getId)
+        Future.successful(())
+      case TASK_RUNNING if !task.hasStartedAt => // was staged, is now running
+        updateTask(task.toBuilder.setStartedAt(System.currentTimeMillis).setStatus(status).build())
+      case _ =>
+        updateTaskOnStateChange(task)
+    }
+  }
+
+  /**
+    * Remove the task for the given app with the given ID completely.
+    *
+    * If the task does not exist, the method will not fail.
+    */
+  override def terminated(appId: PathId, taskId: String): Future[_] = {
+    val app = getInternalApp(appId)
+    removeTask(app, taskId)
+    Future.successful(())
+  }
+
+  private[this] def removeTask(app: InternalApp, taskId: String): Unit = {
+    app.tasks.remove(taskId)
+
+    // FIXME: Should use Future as method return value instead of Await
+    Await.result(repo.expunge(taskId), timeout)
+
+    log.info(s"Task $taskId expunged and removed from TaskTracker")
+
+    if (app.shutdown && app.tasks.isEmpty) {
+      // Are we shutting down this app? If so, remove it
+      remove(app.appName)
+    }
+  }
+
+  override def removeUnknownAppAndItsTasks(appId: PathId): Unit = {
+    val app = getInternalApp(appId)
+    app.shutdown = true
+    if (app.tasks.isEmpty) remove(appId)
+  }
+
+  private[this] def remove(appId: PathId): Unit = {
+    cachedApps.remove(appId)
+    log.info(s"App $appId removed from TaskTracker")
+  }
+
+  // FIXME: See deficiencies above, should probably be replaced by preloading all tasks on elected
+  private[tasks] def fetchApp(appId: PathId): InternalApp = {
+    def fetchTask(taskId: String): Option[MarathonTask] = Await.result(repo.task(taskId), timeout)
```

Comment:
The only difference between this class and the original TaskTracker is that fetchTask was moved to inside this method?